partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt

```diff
@@ -63,32 +64,114 @@ internal class ExprCallDynamic(
             }.toTypedArray()
             return fn.invoke(args)
         }
+    }
+
+    private sealed interface CandidateIndex {
+
+        public fun get(args: List<PartiQLValueType>): Candidate?
+
+        /**
+         * Preserves the original ordering of the passed-in candidates while making it faster to lookup matching
+         * functions. Utilizes both [Direct] and [Indirect].
+         *
+         * @param candidates
+         */
+        class All(
+            candidates: List<Candidate>,
+        ) : CandidateIndex {
+
+            private val lookups: List<CandidateIndex>
+            internal val name: String = candidates.first().fn.signature.name
 
-        internal fun matches(inputs: Array<PartiQLValue>): Boolean {
-            for (i in inputs.indices) {
-                val inputType = inputs[i].type
-                val parameterType = signatureParameters[i]
-                val c = coercions[i]
-                when (c) {
-                    // coercion might be null if one of the following is true
-                    // Function parameter is ANY,
-                    // Input type is null
-                    // input type is the same as function parameter
-                    null -> {
-                        if (!(inputType == parameterType || inputType == PartiQLValueType.NULL || parameterType == PartiQLValueType.ANY)) {
-                            return false
+            init {
```

Comment:
Curious if we see any notable CPU utilization jump from this `init` block since there is more work when an `ExprCallDynamic` is initialized. Wondering if there are any situations in which this construction of the `CandidateIndex` results in worse evaluation performance/utilization (e.g. if `candidates` is just a list of two elements)? I assume this change is better in the general case in which there are more than just a few candidate elements though.

Reply:
Good call out -- in the future, I'd say it'd be a good idea for the `Compiler` to dictate what version of `ExprCallDynamic` to create (the indexing O(1) vs iteration O(n)) depending on the number of candidates (and our own testing). As for `init` by itself, it is only invoked at compilation.

For now, we know dynamic is quite expensive for our customers.